### PR TITLE
refactor: python implementation of `audioop.mul`

### PR DESCRIPTION
**Pull request by davidhozic**
## Summary
This PR replaces the ``PCMVolumeTransformer.read()`` method with a Python equivalent, that does not use the ``audioop`` module, which is deprecated since Python 3.11 and will be removed in Python 3.13.

This implements #2177.

## Information

- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.
- [x] I have updated the changelog to include these changes.

### CODE REVIEW

**Review by VincentRPS — DISMISSED** [review]
LGTM

**Comment by Om1609** [comment]
A python equivalent will not be nearly as close as the C one in terms of performance. Live audio is a heavy load and has to be low latency.

**Comment by VincentRPS** [comment]
> A python equivalent will not be nearly as close as the C one in terms of performance. Live audio is a heavy load and has to be low latency.

I mean we can always rewrite it in Rust.

**Comment by VincentRPS** [comment]
and also this largely uses python stdlib modules which from the seems of things are written in C in many places.

**Comment by Om1609** [comment]
> > A python equivalent will not be nearly as close as the C one in terms of performance. Live audio is a heavy load and has to be low latency.
> 
> I mean we can always rewrite it in Rust.

The whole voice stuff could be rewritten to Rust 😬

**Comment by Om1609** [comment]
> and also this largely uses python stdlib modules which from the seems of things are written in C in many places.

Possible to do actual performance tests?

**Comment by VincentRPS** [comment]
> > > A python equivalent will not be nearly as close as the C one in terms of performance. Live audio is a heavy load and has to be low latency.
> > 
> > 
> > I mean we can always rewrite it in Rust.
> 
> The whole voice stuff could be rewritten to Rust 😬

only if :)

**Comment by davidhozic** [comment]
audioop:
    image

Without audioop:
image

Hardware:
OS Name	Microsoft Windows 10 Pro
Processor	AMD Ryzen 5 5600X 6-Core Processor, 3701 Mhz, 6 Core(s), 12 Logical Processor(s)

**Comment by Om1609** [comment]
Hmm that sure is a difference.

Use `time.perf_counter_ns()` instead of `time.time()`. The result will be more accurate and will return ns instead of s

**Comment by Om1609** [comment]
Can we maintain and ship our own audioop C file? Dpy plans to do something like that

Or we could try Cython or maybe Rust?

**Comment by davidhozic** [comment]
There, it's faster now:
image

**Comment by davidhozic** [comment]
With time.perf_counter_ns():

image

**Comment by davidhozic** [comment]
```py
def timeit(f: Callable, ):
    def wrapper(*args, **kwargs):
        start = time.perf_counter_ns()
        result = f(*args, **kwargs)
        print(f"Elapsed: {(time.perf_counter_ns() - start)} ns")
        return result

    return wrapper
```

for anyone trying to benchmark:

image

**Comment by davidhozic** [comment]
I can rewrite it in C if you'd like, possible compile it first with cython and then optimize?

**Comment by davidhozic** [comment]
CPU running at 1 GHz, latest changes:
image
I think this should be more than enough optimized, since this function reads 20ms of audio.

File: CHANGELOG.md [hunk 1]
```diff
@@ -75,6 +75,9 @@ These changes are available on the `master` branch, but have not yet been releas
 
 ### Changed
 
+- Replaced audioop (deprecated module) implementation of `PCMVolumeTransformer.read`
```
Comment (Lulalaby) [inline CHANGELOG.md:78]: Move to bottom. New entries are appended

**Review by Lulalaby — CHANGES_REQUESTED** [review]

**Comment by davidhozic** [comment]
R5 5600x 4.7 GHz
```
Elapsed: 31819500 ns
Elapsed: 1016100 ns
Elapsed: 868800 ns
Elapsed: 739800 ns
Elapsed: 897400 ns
Elapsed: 743300 ns
Elapsed: 732200 ns
Elapsed: 784500 ns
Elapsed: 840000 ns
Elapsed: 936100 ns
Elapsed: 909200 ns
Elapsed: 731400 ns
Elapsed: 887800 ns
Elapsed: 941000 ns
Elapsed: 950500 ns
Elapsed: 772200 ns
Elapsed: 922500 ns
Elapsed: 942000 ns
Elapsed: 1004100 ns
Elapsed: 964600 ns
Elapsed: 905800 ns
Elapsed: 1057200 ns
Elapsed: 936400 ns
Elapsed: 873700 ns
Elapsed: 1033400 ns
Elapsed: 887700 ns
Elapsed: 953700 ns
Elapsed: 946600 ns
Elapsed: 980300 ns
Elapsed: 797400 ns
Elapsed: 890200 ns
Elapsed: 1122800 ns
Elapsed: 1002900 ns
Elapsed: 939600 ns
Elapsed: 746800 ns
Elapsed: 747200 ns
Elapsed: 1389700 ns
Elapsed: 767300 ns
Elapsed: 950200 ns
Elapsed: 1132400 ns
Elapsed: 766400 ns
Elapsed: 995400 ns
Elapsed: 998400 ns
```

**Comment by Om1609** [comment]
Wow that is some real improvement 👍 

LGTM

File: discord/player.py [hunk 2]
```diff
@@ -652,6 +653,11 @@ class PCMVolumeTransformer(AudioSource, Generic[AT]):
     This does not work on audio sources that have :meth:`AudioSource.is_opus`
     set to ``True``.
 
+    .. versionchanged:: 2.5
+
+        Replaced audioop implementation of :py:meth:`discord.PCMVolumeTransformer.read`
+        method with a pure Python equivalent.
+
```
Comment (Dorukyum) [inline discord/player.py:660]: No need to mention this here, it's an internal code change.
```suggestion
```

**Review by Dorukyum — DISMISSED** [review]

**Comment by davidhozic** [comment]
bump

**Review by Lulalaby — CHANGES_REQUESTED** [review]
Where does this come from

File: CHANGELOG.md [hunk 3]
```diff
@@ -52,6 +52,11 @@ These changes are available on the `master` branch, but have not yet been releas
   ([#2496](https://github.com/Pycord-Development/pycord/pull/2496))
 - ⚠️ **Removed support for Python 3.8.**
   ([#2521](https://github.com/Pycord-Development/pycord/pull/2521))
+- Changed the error message that appears when attempting to add a subcommand group to a
+  subcommand group. ([#2275](https://github.com/Pycord-Development/pycord/pull/2275))
```
Comment (Lulalaby) [inline CHANGELOG.md:56]:
```suggestion
```

**Review by Lulalaby — CHANGES_REQUESTED** [review]

**Comment by davidhozic** [comment]
> Where does this come from

The changelog thing? It was probably added through some weird auto-merge.

**Review by Lulalaby — APPROVED** [review]

**Review by plun1331 — APPROVED** [review]

**Comment by davidhozic** [comment]
Are there any plans or are the main maintainters open for rewriting some of the audio functionalities in Rust (for performance reasons)?

**Comment by DA-344** [comment]
> Are there any plans or are the main maintainters open for rewriting some of the audio functionalities in Rust (for performance reasons)?

Py-cord is meant to be used in (almost) all machines, and writing Rust implementations will prevent some machines to use it anymore because they cannot compile such code.

**Pull request merged**